state/protocol/protocol_state/epochs/fallback_statemachine.go

```diff
@@ -18,16 +25,39 @@ var _ StateMachine = (*FallbackStateMachine)(nil)
 
 // NewFallbackStateMachine constructs a state machine for epoch fallback, it automatically sets
 // InvalidEpochTransitionAttempted to true, thereby recording that we have entered epoch fallback mode.
-func NewFallbackStateMachine(view uint64, parentState *flow.RichProtocolStateEntry) *FallbackStateMachine {
+// No errors are expected during normal operations.
+func NewFallbackStateMachine(params protocol.GlobalParams, view uint64, parentState *flow.RichProtocolStateEntry) (*FallbackStateMachine, error) {
 	state := parentState.ProtocolStateEntry.Copy()
-	state.InvalidEpochTransitionAttempted = true
+	nextEpochCommitted := state.EpochPhase() == flow.EpochPhaseCommitted
+	// we are entering fallback mode, this logic needs to be executed only once
+	if !state.InvalidEpochTransitionAttempted {
+		// the next epoch has not been committed, but possibly setup, make sure it is cleared
+		if !nextEpochCommitted {
+			state.NextEpoch = nil
+		}
+		state.InvalidEpochTransitionAttempted = true
+	}
+
+	if !nextEpochCommitted && view+params.EpochCommitSafetyThreshold() >= parentState.CurrentEpochFinalView() {
```

Comment:
🤔 I have a slightly different view on this. While I agree with Jordan's argument for `OrthogonalStoreStateMachine`, the `FallbackStateMachine` is one level lower. It does _not_ have an `EvolveState` method but ingests the events individually via dedicated methods. Therefore, I see no design inconsistency. Overall, I think it acceptable for package internal sub-components to vary the API according to their needs. Furthermore, we separated the happy path from the fallback path specifically to keep the intellectual complexity manageable and explicitly express the concept of "throw all progress away from the happy path and re-do the evolution using the fallback logic". If we now merge the happy path and the fallback path, I expect this problem to come back. Honestly, I consider code that is hard to reason about way more problematic than a package-internal API convention.

Nevertheless, I don't want to entirely preclude the possibility of merging happy path and fallback again. But I think to make an educated decision, we need the _entire_ logic fully implemented to better understand how similar / different the two paths are.

Reply:
> It does not have an EvolveState method but ingests the events individually via dedicated methods.

The `FallbackStateMachine` itself does not have an `EvolveState` method, but the higher-level `EpochStateMachine` _does_, and it uses the `FallbackStateMachine`. The problem is precisely that we _do not_ have dedicated methods for EFM related state changes, so they go in the constructor.

My point is, the interface suggests that state changes should happen in `EvolveState`. But if I instantiate an instance of `EpochStateMachine`, it might evolve its internal state before I ever call `EvolveState`.

Reply:
>The problem is precisely that we do not have dedicated methods for EFM related state changes, so they go in the constructor.
>My point is, the interface suggests that state changes should happen in EvolveState. But if I instantiate an instance of EpochStateMachine, it might evolve its internal state before I ever call EvolveState.

I don't quite agree with this mindset, I will try to explain why. The idea is that when we create a fallback state machine it means one of two things:
1. we are entering EFM
2. we are already in EFM

In either of those cases we can conclude that EFM is taking place. Constructor of `FallbackStateMachine` takes responsibility of creating a _valid_ instance of state machine that ensures that updated state holds its invariant right after the construction. Without a need to call extra methods. 

Right now this works quite well with how EFM is activated but I am open to making changes to the whole architecture and possibly merging everything together once we see how current design looks after finishing implementation of missing pieces.

Reply:
We still disagree, but let's pick this discussion up again later, if needed, once the state machine changes are more fleshed out. Added a note to [Open Questions](https://www.notion.so/flowfoundation/Spork-less-Epoch-Fallback-Recovery-Design-II-Epoch-Extensions-e923ca84b4d848c5a17de4b18fb5afd5?pvs=4#f1abed16a7784842b6aa35ef2b3c7ee0).